python/ql/src/semmle/python/dataflow/Implementation.qll

```diff
@@ -0,0 +1,849 @@
+
+    /** Hold if taint flows to `src` to `(node, context, path, kind)` in a single step, labelled with `egdeLabel` with this configuration.
+     * `edgeLabel` is purely informative.
+     */
+    predicate flowStep(TaintTrackingNode src, DataFlow::Node node, TaintTrackingContext context, AttributePath path, TaintKind kind, string edgeLabel) {
+        this.unprunedStep(src, node, context, path, kind, edgeLabel) and
+        node.getBasicBlock().likelyReachable() and
+        not this.(TaintTracking::Configuration).isBarrier(node) and
+        (
+            not path = TNoAttribute()
+            or
+            not this.(TaintTracking::Configuration).isBarrier(node, kind) and
+            exists(DataFlow::Node srcnode, TaintKind srckind |
+                src = TTaintTrackingNode_(srcnode, _, _, srckind, this) and
+                not this.prunedEdge(srcnode, node, srckind, kind)
+            )
+        )
+    }
+
+    private predicate prunedEdge(DataFlow::Node srcnode, DataFlow::Node destnode, TaintKind srckind, TaintKind destkind) {
+        this.(TaintTracking::Configuration).isBarrierEdge(srcnode, destnode, srckind, destkind)
+        or
+        srckind = destkind and this.(TaintTracking::Configuration).isBarrierEdge(srcnode, destnode)
+    }
+
+    private predicate unprunedStep(TaintTrackingNode src, DataFlow::Node node, TaintTrackingContext context, AttributePath path, TaintKind kind, string edgeLabel) {
+        this.importStep(src, node, context, path, kind, edgeLabel)
+        or
+        this.fromImportStep(src, node, context, path, kind, edgeLabel)
+        or
+        this.attributeLoadStep(src, node, context, path, kind, edgeLabel)
+        or
+        this.getattrStep(src, node, context, path, kind, edgeLabel)
+        or
+        this.useStep(src, node, context, path, kind, edgeLabel)
+        or
+        this.callTaintStep(src, node, context, path, kind, edgeLabel)
+        or
+        this.returnFlowStep(src, node, context, path, kind, edgeLabel)
+        or
+        this.callFlowStep(src, node, context, path, kind, edgeLabel)
+        or
+        this.iterationStep(src, node, context, path, kind, edgeLabel)
+        or
+        this.yieldStep(src, node, context, path, kind, edgeLabel)
+        or
+        this.parameterStep(src, node, context, path, kind, edgeLabel)
+        or
+        this.ifExpStep(src, node, context, path, kind, edgeLabel)
+        or
+        this.essaFlowStep(src, node, context, path, kind, edgeLabel)
+        or
+        this.instantiationStep(src, node, context, path, kind, edgeLabel)
+        or
+        this.legacyExtensionStep(src, node, context, path, kind, edgeLabel)
+        or
+        exists(DataFlow::Node srcnode, TaintKind srckind |
+            this.(TaintTracking::Configuration).isAdditionalFlowStep(srcnode, node, srckind, kind) and
+            src = TTaintTrackingNode_(srcnode, context, path, srckind, this) and
+            path.noAttribute() and edgeLabel = "additional with kind"
```

Comment:
Would it make sense to add the kind to the label? I'm thinking that for debugging purposes, it might be useful to know which extension was responsible. (Although... I guess with configurations, there should rarely be much confusion about this.)

Reply:
We want to keep the overhead of these labels to a minimum, so they should only be constants.